CMakeLists.txt

```diff
@@ -1,367 +1,165 @@
-
-set(INCLUDE_INSTALL_DIR ${INCLUDE_INSTALL_ROOT_DIR}/yaml-cpp)
-set(LIB_INSTALL_DIR "lib${LIB_SUFFIX}")
-
-set(_INSTALL_DESTINATIONS
-	RUNTIME DESTINATION bin
-	LIBRARY DESTINATION ${LIB_INSTALL_DIR}
-	ARCHIVE DESTINATION ${LIB_INSTALL_DIR}
-)
-
-
-###
-### Library
-###
-if(YAML_BUILD_SHARED_LIBS)
-    add_library(yaml-cpp SHARED ${library_sources})
-else()
-    add_library(yaml-cpp STATIC ${library_sources})
-endif()
-
-if (NOT CMAKE_VERSION VERSION_LESS 2.8.12)
-    target_include_directories(yaml-cpp
-        PUBLIC $<BUILD_INTERFACE:${YAML_CPP_SOURCE_DIR}/include>
-               $<INSTALL_INTERFACE:${INCLUDE_INSTALL_ROOT_DIR}>
-        PRIVATE $<BUILD_INTERFACE:${YAML_CPP_SOURCE_DIR}/src>)
-endif()
-
-set_target_properties(yaml-cpp PROPERTIES
-    CXX_STANDARD 11
-    CXX_STANDARD_REQUIRED ON
-)
-
-set_target_properties(yaml-cpp PROPERTIES
-  COMPILE_FLAGS "${yaml_c_flags} ${yaml_cxx_flags}"
-)
+cmake_dependent_option(YAML_CPP_BUILD_TESTS
+  "Enable yaml-cpp tests" ON
+  "BUILD_TESTING;CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
+cmake_dependent_option(YAML_CPP_INSTALL
+  "Enable generation of yaml-cpp install targets" ON
+  "CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR" OFF)
+cmake_dependent_option(YAML_BUILD_SHARED_LIBS
+  "Build yaml-cpp shared library" OFF
+  "BUILD_SHARED_LIBS" OFF)
+
+cmake_dependent_option(YAML_MSVC_SHARED_RT
+  "MSVC: Build yaml-cpp with shared runtime libs (/MD)" ON
+  "MSVC" OFF)
+cmake_dependent_option(YAML_APPLE_UNIVERSAL_BIN
+  "Apple: Build yaml-cpp universal binary" OFF
+  "APPLE" OFF)
+
+set(yaml-cpp-type STATIC)
+if (YAML_BUILD_SHARED_LIBS)
+  set(yaml-cpp-type SHARED)
+endif()
+
+set(build-shared $<BOOL:${YAML_BUILD_SHARED_LIBS}>)
+set(build-windows-dll $<AND:$<BOOL:${CMAKE_HOST_WIN32}>,${build-shared}>)
+set(not-msvc $<NOT:$<CXX_COMPILER_ID:MSVC>>)
+
+if (NOT DEFINED CMAKE_MSVC_RUNTIME_LIBRARY)
+  set(CMAKE_MSVC_RUNTIME_LIBRARY
+    MultiThreaded$<$<CONFIG:Debug>:Debug>$<${build-shared}:DLL>)
+endif()
+
+set(contrib-pattern "src/contrib/*.cpp")
+set(src-pattern "src/*.cpp")
+if (CMAKE_VERSION VERSION_GREATER 3.12)
+  list(INSERT contrib-pattern 0 CONFIGURE_DEPENDS)
+  list(INSERT src-pattern 0 CONFIGURE_DEPENDS)
+endif()
+
+file(GLOB yaml-cpp-contrib-sources ${contrib-pattern})
+file(GLOB yaml-cpp-sources ${src-pattern})
+
+# /wd4127 = disable warning C4127 "conditional expression is constant"
+# http://msdn.microsoft.com/en-us/library/6t66728h.aspx
+# /wd4355 = disable warning C4355 "'this' : used in base member initializer list
+# http://msdn.microsoft.com/en-us/library/3c594ae3.aspx
+set(msvc-rt $<TARGET_PROPERTY:MSVC_RUNTIME_LIBRARY>)
+
+set(mtd-static $<STREQUAL:${msvc-rt},MultiThreadedDebug>)
+set(mt-static $<STREQUAL:${msvc-rt},MultiThreaded>)
+
+set(mtd-dll $<STREQUAL:${msvc-rt},MultiThreadedDebugDLL>)
+set(mt-dll $<STREQUAL:${msvc-rt},MultiThreadedDLL>)
+
+set(backport $<VERSION_LESS:${CMAKE_VERSION},3.15>)
+
+add_library(yaml-cpp ${yaml-cpp-type})
+add_library(yaml::yaml ALIAS yaml-cpp)
```

Comment:
What does this mean?

Reply:
When `yaml-cpp` is used via `add_subdirectory`, users can link against `yaml::yaml` and anyone who might pull in their project will not be able to modify the properties of it. ALIAS libraries are explained in detail [here][1]

[1]: https://cmake.org/cmake/help/latest/command/add_library.html#alias-libraries

Reply:
So is the idea that the yaml-cpp target is only internally visible, and yaml::yaml is externally visible, and immutable?

If so, is yaml-cpp really only internally visible, and if so, what makes it so?

Reply:
`yaml-cpp` isn't internally visible if used in an `add_subdirectory` form, but *can* be partially internally visible if its an `IMPORTED` library, and not marked as `GLOBAL` (which is the default in `find_package`)

IIRC I fixed (or have a todo to fix) the `install(TARGETS ... EXPORT ...)`  `find_package` file to make sure `yaml::yaml` is available. This means that the actual exported target name doesn't really matter, as an ALIAS is used instead, and in the case of `find_package`, one wouldn't be able to modify it. Doing it this way allows people to swap between an installed version, or one they pull in manually, without having upstream projects having to check or care.

The scoping of libraries in CMake is weird, and we unfortunately can't change Kitware's minds on how to fix them because that would require breaking backwards compatibility. 

It's also hella confusing, so I'm trying to reduce that while giving upstream users the option.